### ISSUE TITLE
Stage to check and supply Jenkins slaves etc.

### ISSUE THREAD

**Issue by michaelsauter** [opendevstack/ods-quickstarters#14]
Handle common dependencies of quickstarters (e.g. build images, slaves, ...). Right now each quickstarter assumes that e.g. the Jenkins slave magically exists. If we want to have a dynamic number of quickstarte (repos), we probably need some "register"/"install" function to make sure a quickstarter can be used

**Comment by gerardcl** [comment]
With @velcrin we have been discussing if slave base childs are really required indeed. 
Pros:
- specific configurations and integrations with ODS and OpenShift itself
- others?

Cons:
- tech debt
- this issue

Proposal:
- make an implicit use of the kubernetes jenkins plugin by exposing the podContainers descriptor when specific techs are required
- so to always (at least) make use of the slave-base-image for the Jenkins master<->slave architecture and ODS/OpenShift integrations (tailor, oc, git lfs, sonarqube, ...)
- only load specific techs based on the quickstarter needs (based on official images of that specific tech)

**Comment by michaelsauter** [comment]
@gerardcl Sorry, I'm not sure I'm following. What exactly are you proposing? Could you explain with an example maybe?

**Comment by gerardcl** [comment]
sure!
basically the proposal is to get rid (or try to get fully rid of) the slave base childs (e.g.: jenkins-slave-nodejs)

and this could happen since for stages where binaries are required we can keep using the jenkins-slave-base (stages like sonarqube scan, deploy and build to openshift) and then for build, linting, testing, we could just make use of the official images (e.g.: image node:10 for the pipelines that require npm and so on)

so an example change for example for the be-python-flask could be:

```
- image: "${dockerRegistry}/cd/jenkins-slave-python:@ods_image_tag@"
+   podContainers: [
    containerTemplate(
      name: 'jnlp',
      image: "${dockerRegistry}/cd/jenkins-slave-base:@ods_image_tag@",
      workingDir: '/tmp',
      alwaysPullImage: true,
      args: '${computer.jnlpmac} ${computer.name}',
      serviceAccount: 'jenkins'
    ),
    containerTemplate(
      name: 'python',
      image: "python:3.6-buster",
      workingDir: '/tmp',
      alwaysPullImage: true,
      ttyEnabled: true,
      command: 'cat',
    )
  ],
```

and then just use the DSL statement:
```
container("python") {
  stageBuild()
  stageTest()
}
```
wherever only python tools are required

**Comment by michaelsauter** [comment]
Got it! Thanks for the explanation.

That sounds quite interesting! It could also solve issues in differing architecture between build-time image and runtime image (as we have seen with Python, I think @adyachok and @henrjk have encountered issues there). Though we can also solve those issues by going to UBI images, your suggestion still has lots of appeal.

@gerardcl @adyachok @henrjk @velcrin  Could you guys, since you have already dealt with this a bit, give this approach a shot and see if it is viable for a handful of quickstarters? That would be awesome. The smallest MVP might be two use a different key, such as `containerImage`, which allows user to set the "specific image"?

**Comment by henrjk** [comment]
Currently I don't have time to actively pursue this (reviewing should be ok).
But I think this is the right direction as in Python with native packages it does not really make sense to test in a container of a different breed than what is later running. Of course it would be a bonus to arrange it so to have the runtime image leaner than what is used for building. However I would consider this out of scope unless there are some low hanging fruit.

**Comment by michaelsauter** [comment]
I just realised that doing this change would increase resource usage a lot. That's a real bummer ... not sure if the benefits are big enough to afford this

**Comment by gerardcl** [comment]
Yes, true! But: I think that on average there will be a maximum of two containers per quickstarter. Which we can limit them by the half so to keep same current resources consumption, if possible.
Anyway, users might require to extend with other containers too, which might lead to require from their side fine-tunning the resources their request or request to cluster admins for exceptions and increase limits defined, if quotas for example.
Shall we give it a try or we close it? I might try to explore this with @velcrin in the following weeks whenever we have time, so to test with the python qs and some javascript related ones?
Do you have any more concerns?
This can go anyway for ODSv4...

**Comment by michaelsauter** [comment]
Yes, I actually have more concerns:

* I'm wondering - if you build your app one one container, but the OpenShift container build is started from another one, you have to transfer the artefacts between the containers.
* RedHat seems to suggest the "one slave" architecture by providing a Maven slave and a Node slave

I am really not sure we should follow this ... but feel free to experiment!
